### PR TITLE
8289301: P11Cipher should not throw out of bounds exception during padding

### PR DESCRIPTION
**Pull request by zzambers**
SunPkcs11 provider throws out of bounds exception during encryption when specific conditions are met.

Exception:
```
Exception in thread "main" java.lang.ArrayIndexOutOfBoundsException: Array index out of range: 32
	at java.base/java.util.Arrays.rangeCheck(Arrays.java:725)
	at java.base/java.util.Arrays.fill(Arrays.java:3308)
	at jdk.crypto.cryptoki/sun.security.pkcs11.P11Cipher$PKCS5Padding.setPaddingBytes(P11Cipher.java:96)
	at jdk.crypto.cryptoki/sun.security.pkcs11.P11Cipher.implDoFinal(P11Cipher.java:813)
	at jdk.crypto.cryptoki/sun.security.pkcs11.P11Cipher.engineDoFinal(P11Cipher.java:585)
	at java.base/javax.crypto.Cipher.doFinal(Cipher.java:2417)
...
```

Details:
This problems happens when reqBlockUpdates is true and implUpdate, which does not end on block boundary, is performed followed by final implUpdate, which ends exactly on block boundary. In that case final implUpdate fills padBuffer and then just returns. [1] Following implDoFinal then tries to add padding and throws OOB exception. Problem is, that in this case (input is multiple of block size) whole padding block should be added, but there is no space for it in padBuffer causing OOB exception.

Solution:
Solution is to detect this case (implDoFinal is called with full padBuffer) and to perform additional C_EncryptUpdate to free up padBuffer so that padBuffer can than be used to add whole new padding block.

[1] https://github.com/openjdk/jdk/blob/d4eeeb82cb2288973a6a247c54513f7e1c6b58f0/src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/P11Cipher.java#L622

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8289301](https://bugs.openjdk.org/browse/JDK-8289301): P11Cipher should not throw out of bounds exception during padding

### Reviewers
 * [Valerie Peng](https://openjdk.org/census#valeriep) (@valeriepeng - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/9310/head:pull/9310` \
`$ git checkout pull/9310`

Update a local copy of the PR: \
`$ git checkout pull/9310` \
`$ git pull https://git.openjdk.org/jdk pull/9310/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 9310`

View PR using the GUI difftool: \
`$ git pr show -t 9310`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/9310.diff">https://git.openjdk.org/jdk/pull/9310.diff</a>

</details>

### CODE REVIEW

**Comment by zzambers** [comment]
jdk_security tests passed for me locally

**Comment by valeriepeng** [comment]
I am still reviewing some other PR, will take a look of this tomorrow. Thanks~

File: src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/P11Cipher.java [hunk 1]
```diff
@@ -814,13 +814,23 @@ private int implDoFinal(byte[] out, int outOfs, int outLen)
                 if (paddingObj != null) {
                     int startOff = 0;
                     if (reqBlockUpdates) {
-                        startOff = padBufferLen;
+                        // padding is added to data in padBuffer
+                        // unless padBuffer is already full in which case
+                        // EncryptUpdate is performed and whole new padding
+                        // block is created
+                        if (padBufferLen == padBuffer.length) {
+                            k += token.p11.C_EncryptUpdate(session.id(),
```
Comment (valeriepeng) [inline src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/P11Cipher.java:822]: can just be "="?

File: src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/P11Cipher.java [hunk 2]
```diff
@@ -814,13 +814,23 @@ private int implDoFinal(byte[] out, int outOfs, int outLen)
                 if (paddingObj != null) {
                     int startOff = 0;
                     if (reqBlockUpdates) {
-                        startOff = padBufferLen;
+                        // padding is added to data in padBuffer
+                        // unless padBuffer is already full in which case
+                        // EncryptUpdate is performed and whole new padding
+                        // block is created
```
Comment (valeriepeng) [inline src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/P11Cipher.java:820]: suggestion on comment: 
// call C_EncryptUpdate first if the padBuffer is full to make room for padding bytes

File: src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/P11Cipher.java [hunk 3]
```diff
@@ -902,13 +912,23 @@ private int implDoFinal(ByteBuffer outBuffer)
...
```
Comment (valeriepeng) [inline src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/P11Cipher.java:918]: Same suggestion on this as above.

File: test/jdk/sun/security/pkcs11/Cipher/TestPaddingOOB.java [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+/*
+ * @test
+ * @bug 8289301
+ * @summary P11Cipher should not throw OOB exception during padding
```
Comment (valeriepeng) [inline test/jdk/sun/security/pkcs11/Cipher/TestPaddingOOB.java:27]: nit: add when "reqBlockUpdates" == true

File: test/jdk/sun/security/pkcs11/Cipher/TestPaddingOOB.java [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+        int off = c.update(plainArr, 0, 1, encArr, 0);
+        off += c.doFinal(plainArr, 1, plainArr.length - 1, encArr, off);
+        if (off != 2 * bs) {
+            throw new Exception("Unexpected encrypted size (array)");
+        }
+        c.init(Cipher.DECRYPT_MODE, key);
+        byte[] plainArr2 = new byte[c.getOutputSize(encArr.length)];
+        off = c.doFinal(encArr, 0, encArr.length, plainArr2, 0);
+        if (off != bs) {
+            throw new Exception("Unexpected decrypted size (array)");
```
Comment (valeriepeng) [inline test/jdk/sun/security/pkcs11/Cipher/TestPaddingOOB.java:67]: nit: add the value of "off" to the exception message which may come in handy if the test were to fail later.

**Comment by zzambers** [comment]
I have fixed all the issues, jdk_security tests passed for me (with these changes). Thanks

**Comment by valeriepeng** [comment]
> I have fixed all the issues, jdk_security tests passed for me (with these changes). Thanks

Looks good, I will approve after the test job complete successfully. Thanks!

**Review by valeriepeng — APPROVED** [review]

**Comment by zzambers** [comment]
@valeriepeng Thank you for the review
/integrate

**Pull request closed without merging**